src/connection.rs

```diff
@@ -286,3 +291,142 @@ where
         }
     }
 }
+
+pub struct Select<'a> {
+    cursor: Option<Cursor<'a>>,
+    error: Option<Error>,
+    columns_map: HashMap<String, usize>,
+}
+
+impl<'a> Select<'a> {
+    pub fn query(conn: &'a Connection, query: impl AsRef<str>) -> Self {
+        match conn.prepare(query) {
+            Ok(statement) => Self {
+                error: None,
+                columns_map: Self::create_columns_map(&statement),
+                cursor: Some(statement.into_cursor()),
+            },
+            Err(err) => Self {
+                cursor: None,
+                error: Some(err),
+                columns_map: Default::default(),
+            },
+        }
+    }
+
+    fn create_columns_map(statement: &Statement) -> HashMap<String, usize> {
+        (0..statement.column_count())
+            .map(|i| (statement.column_name(i).to_string(), i))
+            .collect()
+    }
+}
+
+impl<'a> From<Statement<'a>> for Select<'a> {
+    fn from(statement: Statement<'a>) -> Self {
+        Self {
+            error: None,
+            columns_map: Self::create_columns_map(&statement),
+            cursor: Some(statement.into_cursor()),
+        }
+    }
+}
+
+impl<'a> Iterator for Select<'a> {
+    type Item = Result<Row>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let columns_map = self.columns_map.clone();
+        match (self.cursor.as_mut(), self.error.take()) {
+            (_, Some(err)) => Some(Err(err)),
+            (Some(cursor), _) => match cursor.next() {
+                Err(err) => Some(Err(err)),
+                Ok(maybe_row) => Ok(maybe_row.map(|row| Row {
+                    row: row.to_vec(),
+                    columns_map,
+                }))
+                .transpose(),
+            },
+            (None, None) => None,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct Row {
+    row: Vec<Value>,
+    columns_map: HashMap<String, usize>,
+}
+
+pub trait ValueInto: Sized {
```

Comment:
Not sure if this trait is needed. Maybe it can be combined with `Value`? Like `impl Into for Value`. If indeed necessary, let’s call the method simply `into`.

Reply:
Yes it is needed because it is not possible to transform this to `TryInto`:

```rust
impl<T: ValueInto> ValueInto for Option<T> {
    fn try_convert_value_into(value: &Value) -> Option<Self> {
        match value {
            Value::Null => Some(None),
            _ => T::try_convert_value_into(value).map(Some),
        }
    }
}
```

Would be:

```rust
impl<T> std::convert::TryInto<Option<T>> for &Value {
    type Error = Error;

    fn try_into(self) -> Result<Option<T>> {
        todo!()
    }
}
```

Which fails to compile because of the lack of specialization I assume:

```
error[E0119]: conflicting implementations of trait `std::convert::TryInto<std::option::Option<_>>` for type `&Value`
   --> src/lib.rs:332:1
    |
332 | impl<T> convert::TryInto<Option<T>> for &Value {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: conflicting implementation in crate `core`:
            - impl<T, U> TryInto<U> for T
              where U: TryFrom<T>;
```